platform/linux-generic/odp_crypto.c

```diff
@@ -112,8 +116,32 @@ struct odp_crypto_generic_session_t {
 		const EVP_MD *evp_md;
 		crypto_func_t func;
 	} auth;
+
+	/* These bitfields are cleared at odp_crypto_session_destroy()
+	 * together with the rest of data */
+	crypto_valid cipher_valid[(ODP_THREAD_COUNT_MAX + CV_BITS - 1) /
+				CV_BITS];
+	crypto_valid hmac_valid[(ODP_THREAD_COUNT_MAX + CV_BITS - 1) /
+				CV_BITS];
+	unsigned idx;
 };
 
+/* Use _mm versions, because they are always inlined */
+static int crypto_should_init(crypto_valid *ptr, unsigned int id)
+{
+	unsigned int mask = 1 << (id % CV_BITS);
+	crypto_valid *cv = ptr + (id / CV_BITS);
+	uint32_t cur = _odp_atomic_u32_load_mm(ptr, _ODP_MEMMODEL_ACQ);
+
+
+	while (!_odp_atomic_u32_cmp_xchg_strong_mm(cv, &cur, cur | mask,
```

Comment:
I think the atomic cmp-xchg is too heavyweight here.

AFAICS, it is needed only because of the way the valid bits for different threads are packed in the same memory locations. The need for the syncing atomic op could be avoided by just using a whole uint8_t for one thread. I do not think the increase in memory use would be so big that it would be a problem.

If you really want to keep packing the valid bits, then you could do so also by having per-thread arrays of valid bits for all sessions in the global SHM instead of the current approach of per-session arrays of valid bits for all threads. That would avoid the syncing needs since the valid bits would not share memory locations with other threads.

IOW, you would have something like this in the global SHM:

struct openssl_ctx_status {
	crypto_valid cipher_valid[(MAX_SESSIONS + CV_BITS - 1) / CV_BITS];
	crypto_valid   hmac_valid[(MAX_SESSIONS + CV_BITS - 1) / CV_BITS];
} [ODP_THREAD_COUNT_MAX];

Reply:
Hmm. I like this idea. I will implement it, thank you.

Reply:
@JannePeltonen even with your suggestion there can be congestion between set-as-valid session1 and clear-valid session2 (which can happen in completely different thread). In fact I'm balancing between two possible solutions:

- always init all contexts on session init/destroy them on session destroy
- just add a per-thread lock guarding writing to per-thread bits

Any suggestions?